prov/gni/include/gnix_tags.h

```diff
@@ -0,0 +1,416 @@
+ *
+ * This software is available to you under a choice of one of two
+ * licenses.  You may choose to be licensed under the terms of the GNU
+ * General Public License (GPL) Version 2, available from the file
+ * COPYING in the main directory of this source tree, or the
+ * BSD license below:
+ *
+ *     Redistribution and use in source and binary forms, with or
+ *     without modification, are permitted provided that the following
+ *     conditions are met:
+ *
+ *      - Redistributions of source code must retain the above
+ *        copyright notice, this list of conditions and the following
+ *        disclaimer.
+ *
+ *      - Redistributions in binary form must reproduce the above
+ *        copyright notice, this list of conditions and the following
+ *        disclaimer in the documentation and/or other materials
+ *        provided with the distribution.
+ *
+ * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND,
+ * EXPRESS OR IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF
+ * MERCHANTABILITY, FITNESS FOR A PARTICULAR PURPOSE AND
+ * NONINFRINGEMENT. IN NO EVENT SHALL THE AUTHORS OR COPYRIGHT HOLDERS
+ * BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER LIABILITY, WHETHER IN AN
+ * ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM, OUT OF OR IN
+ * CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
+ * SOFTWARE.
+ */
+
+/*
+ * Examples:
+ *
+ * Init:
+ *
+ * _gnix_posted_tag_storage_init(&ep->posted_recvs, NULL);
+ * _gnix_unexpected_tag_storage_init(&ep->unexpected_recvs, NULL);
+ *
+ * On receipt of a message:
+ *
+ * fastlock_acquire(&ep->tag_lock);
+ * req = _gnix_remove_by_tag(&ep->posted_recvs, msg->tag, 0);
+ * if (!req)
+ *     _gnix_insert_by_tag(&ep->unexpected_recvs, msg->tag, msg->req);
+ * fastlock_release(&ep->tag_lock);
+ *
+ * On post of receive:
+ *
+ * fastlock_acquire(&ep->tag_lock);
+ * tag_req = _gnix_remove_by_tag(&ep->unexpected_recvs,
+ *           req->tag, req->ignore);
+ * if (!tag_req)
+ *     _gnix_insert_by_tag(&ep->posted_recvs, tag, req);
+ * fastlock_release(&ep->tag_lock);
+ *
+ */
+
+#ifndef PROV_GNI_SRC_GNIX_TAGS_H_
+#define PROV_GNI_SRC_GNIX_TAGS_H_
+
+#include <fi.h>
+#include <fi_list.h>
+
+#include "gnix_util.h"
+
+/* enumerations */
+
+/**
+ * Enumeration for determining the underlying data structure for a
+ * tag storage.
+ *
+ * Using auto will choose one of list, hlist or kdtree based on mem_tag_format.
+ */
+enum {
+	GNIX_TAG_AUTOSELECT = 0,//!< GNIX_TAG_AUTOSELECT
+	GNIX_TAG_LIST,          //!< GNIX_TAG_LIST
+	GNIX_TAG_HLIST,         //!< GNIX_TAG_HLIST
+	GNIX_TAG_KDTREE,        //!< GNIX_TAG_KDTREE
+	GNIX_TAG_MAXTYPES,      //!< GNIX_TAG_MAXTYPES
+};
+
+/**
+ * Enumeration for the tag storage states
+ */
+enum {
+	GNIX_TS_STATE_UNINITIALIZED = 0,//!< GNIX_TS_STATE_UNINITIALIZED
+	GNIX_TS_STATE_INITIALIZED,      //!< GNIX_TS_STATE_INITIALIZED
+	GNIX_TS_STATE_DESTROYED,        //!< GNIX_TS_STATE_DESTROYED
+};
```

Comment:
We have a lot of extra internal stuff in here, presumably just for testing.  It would be nice to clearly separate the interface from the exposed internals.